configomatic/logging.py

```diff
@@ -21,39 +22,69 @@ class LoggingConfiguration(BaseModel):
     # See https://docs.python.org/3/library/logging.config.html#logging-config-dictschema
     version: int = 1
     disable_existing_loggers: bool = False
-    formatters: dict = Field(default_factory = lambda: {
-        "default": {
-            "format": "[%(levelname)s] %(message)s",
-        },
-    })
-    filters: dict = Field(default_factory = lambda: {
-        # This filter allows us to send >= WARNING to stderr and < WARNING to stdout
-        "less_than_warning": {
-            "()": f"{__name__}.LessThanLevelFilter",
-            "level": "WARNING",
-        },
-    })
-    handlers: dict = Field(default_factory = lambda: {
-        "stdout": {
-            "class": "logging.StreamHandler",
-            "stream": "ext://sys.stdout",
-            "formatter": "default",
-            "filters": ["less_than_warning"],
-        },
-        "stderr": {
-            "class": "logging.StreamHandler",
-            "stream": "ext://sys.stderr",
-            "formatter": "default",
-            "level": "WARNING",
-        },
-    })
-    loggers: dict = Field(default_factory = lambda: {
-        "": {
-            "handlers": ["stdout", "stderr"],
-            "level": "INFO",
-            "propagate": True
-        },
-    })
+    formatters: dict = Field(default_factory = dict)
+    filters: dict = Field(default_factory = dict)
+    handlers: dict = Field(default_factory = dict)
+    loggers: dict = Field(default_factory = dict)
+
+    @validator("formatters", pre = True, always = True)
+    def default_formatters(cls, v):
+        return deep_update(
+            {
+                "default": {
+                    "format": "[%(asctime)s] %(name)-20.20s [%(levelname)-8.8s] %(message)s",
+                },
+            },
+            v or {}
+        )
+
+    @validator("filters", pre = True, always = True)
+    def default_filters(cls, v):
+        return deep_update(
+            {
+                # This filter allows us to send >= WARNING to stderr and < WARNING to stdout
+                "less_than_warning": {
+                    "()": f"{__name__}.LessThanLevelFilter",
+                    "level": "WARNING",
+                },
+            },
+            v or {}
+        )
+
+    @validator("handlers", pre = True, always = True)
+    def default_handlers(cls, v):
+        return deep_update(
+            {
+                # Handlers for stdout/err with default formatting
```

Comment:
Could you explain the need for splitting over stdout/stderr? Could we just send all messages to stderr?

Reply:
I don't really want DEBUG and INFO to go to stderr

Reply:
I guess it is a personal thing. It just means I can filter on stderr in Loki and only see errors and warnings 🤷‍♂️

Reply:
All fair enough